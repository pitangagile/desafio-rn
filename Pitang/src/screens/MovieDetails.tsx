import React, { useState, useEffect, useContext } from 'react';
import { Text, View, StyleSheet, ActivityIndicator } from 'react-native';
import styled from 'styled-components/native';
import Image from 'react-native-fast-image';
import { BlurView } from '@react-native-community/blur';
import { iOSUIKit } from 'react-native-typography';
import MovieService from '../service';
import BackHeader from '../components/BackHeader';
import { Movie } from '../typings';
import MovieContext from '../context/MovieContext';
import sizes from '../helpers/utils';

function LoadingSynopsis() {
  return (
    <>
      <Loader />
      <LoadingText>Loading synopsis...</LoadingText>
    </>
  );
}

const Loader = styled.ActivityIndicator.attrs({
  size: 'large',
  color: 'white',
})``;

const LoadingText = styled.Text`
  align-self: center;
  font-size: 20px;
  color: white;
`;

const size = sizes() === 'small' || sizes() === 'medium';

const MovieDetails = ({ route }: any) => {
  const { NOT_FOUND_IMAGE_URL } = useContext(MovieContext);
  const { item } = route.params;
  const [isLoadingImage, setLoadingImage] = useState(false);
  const [details, setDetails] = useState<Movie>({} as Movie);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    loadDetails();
  }, []);

  async function loadDetails() {
    try {
      const movie = await MovieService.details(item._id);
      setDetails(movie);
      setLoading(false);
    } catch (error) {
      setLoading(false);
    }
  }

  return (
    <Container>
      <Background source={{ uri: item.url }}>
        <BlurView
          style={styles.absolute}
          blurType="dark"
          blurAmount={100}
          reducedTransparencyFallbackColor="white"
        />
        <BackHeader />
        <FastImage
          resizeMode={FastImage.resizeMode.contain}
          fallback
          source={{
            uri: item?.url ? item?.url : NOT_FOUND_IMAGE_URL,
            priority: Image.priority.normal,
          }}
          onLoadStart={() => setLoadingImage(true)}
          onLoadEnd={() => setLoadingImage(false)}
        />
        {isLoadingImage && <ActivityIndicator size={'small'} color={'white'} />}
      </Background>
      <MovieInfoWrapper>
        <View>
          <Text style={iOSUIKit.largeTitleEmphasizedWhite}>{item.name}</Text>
          <Divider />
        </View>
        {loading && <LoadingSynopsis />}
        {!loading && (
          <Synopsis>
            <Text style={iOSUIKit.title3EmphasizedWhite}>Synopsis</Text>
            <Description style={iOSUIKit.bodyWhite}>
              {details.description}
            </Description>
          </Synopsis>
        )}
      </MovieInfoWrapper>
    </Container>
  );
};

const FastImage = styled(Image)`
  width: ${size ? 200 : 300};
  align-self: center;
  height: ${size ? 200 : 300};
`;

const Container = styled.SafeAreaView`
  flex: 1;
  background: transparent;
`;

const MovieInfoWrapper = styled.View`
  flex: 1;
  padding: 15px;
  border-bottom-width: 1;
  background: black;
`;

const Divider = styled.View`
  border-width: 1;
  border-color: rgba(255, 255, 255, 0.1);
  margin-top: 20px;
`;

const Synopsis = styled.ScrollView`
  margin-top: 20px;
`;

const Description = styled.Text`
  letter-spacing: 2;
  margin-top: 20px;
`;

const Background = styled.ImageBackground`
  flex: 1;
  justify-content: center;
  align-content: center;
  align-items: center;
`;

const styles = StyleSheet.create({
  absolute: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
  },
});

export default React.memo(MovieDetails);
