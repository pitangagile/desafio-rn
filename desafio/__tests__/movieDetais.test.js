import React from 'react';
import MovieDetails from '../src/components/movieDetails';

import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react-native';

describe('<MovieDetails />', () => {
  it('MovieDetails Snapshot', () => {
    const fakeData = [
      {
        title: 'Avatar',
        image:
          'https://http2.mlstatic.com/poster-cartaz-avatar-2-40x60cm-D_NQ_NP_701026-MLB25905406913_082017-O.webp',
        description:
          'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie eu velit et consequat. Maecenas a velit nec nibh euismod egestas at et eros. Proin imperdiet laoreet diam, at vehicula lacus ornare et. Proin auctor convallis bibendum.',
      },
    ];
    const fragment = renderer
      .create(
        <MovieDetails
          image={fakeData.image}
          title={fakeData.title}
          description={fakeData.description}
          onPress={() => null}
        />,
      )
      .toJSON();
    expect(fragment).toMatchSnapshot();
  });

  it('MovieDetails Press', () => {
    const fakeData = [
      {
        title: 'Avatar',
        image:
          'https://http2.mlstatic.com/poster-cartaz-avatar-2-40x60cm-D_NQ_NP_701026-MLB25905406913_082017-O.webp',
        description:
          'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie eu velit et consequat. Maecenas a velit nec nibh euismod egestas at et eros. Proin imperdiet laoreet diam, at vehicula lacus ornare et. Proin auctor convallis bibendum.',
      },
    ];
    const pressTouch = jest.fn();

    const { getByTestId } = render(
      <MovieDetails
        image={fakeData.image}
        title={fakeData.title}
        description={fakeData.description}
        onPress={pressTouch}
      />,
    );

    fireEvent.press(getByTestId('back'));
    expect(getByTestId('back')).toBeTruthy();
    expect(pressTouch).toHaveBeenCalledTimes(0);
  });
});
