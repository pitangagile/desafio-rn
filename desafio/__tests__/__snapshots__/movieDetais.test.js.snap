// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieDetails /> MovieDetails Snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="back"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "bold",
          }
        }
      >
        &lt;&lt; Voltar
      </Text>
    </View>
    <View
      style={
        Object {
          "height": 450,
          "marginBottom": 15,
          "marginTop": 15,
          "width": "100%",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": undefined,
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "flex": 1,
            "resizeMode": "contain",
            "width": "100%",
          }
        }
      />
    </View>
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 15,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 16,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;
